cpp/src/arrow/compute/exec/exec_plan.cc

```diff
@@ -565,5 +565,253 @@ AsyncGenerator<util::optional<ExecBatch>> MakeSinkNode(ExecNode* input,
   return out;
 }
 
+struct GroupByNode : ExecNode {
+  GroupByNode(ExecNode* input, std::string label, std::shared_ptr<Schema> output_schema,
+              ExecContext* ctx, const std::vector<int>&& key_field_ids,
+              std::unique_ptr<internal::Grouper>&& grouper,
+              const std::vector<int>&& agg_src_field_ids,
+              const std::vector<const HashAggregateKernel*>&& agg_kernels,
+              std::vector<std::unique_ptr<KernelState>>&& agg_states)
+      : ExecNode(input->plan(), std::move(label), {input}, {"groupby"},
+                 std::move(output_schema), /*num_outputs=*/1),
+        ctx_(ctx),
+        key_field_ids_(std::move(key_field_ids)),
+        grouper_(std::move(grouper)),
+        agg_src_field_ids_(std::move(agg_src_field_ids)),
+        agg_kernels_(std::move(agg_kernels)),
+        agg_states_(std::move(agg_states)) {}
+
+  const char* kind_name() override { return "GroupByNode"; }
+
+  Status ProcessInputBatch(const ExecBatch& batch) {
+    // Create a batch with key columns
+    std::vector<Datum> keys(key_field_ids_.size());
+    for (size_t i = 0; i < key_field_ids_.size(); ++i) {
+      keys[i] = batch.values[key_field_ids_[i]];
+    }
+    ARROW_ASSIGN_OR_RAISE(ExecBatch key_batch, ExecBatch::Make(keys));
+
+    // Create a batch with group ids
+    ARROW_ASSIGN_OR_RAISE(Datum id_batch, grouper_->Consume(key_batch));
+
+    // Execute aggregate kernels
+    auto num_groups = grouper_->num_groups();
+    for (size_t i = 0; i < agg_kernels_.size(); ++i) {
+      KernelContext kernel_ctx{ctx_};
+      kernel_ctx.SetState(agg_states_[i].get());
+      ARROW_ASSIGN_OR_RAISE(
+          auto agg_batch, ExecBatch::Make({batch.values[agg_src_field_ids_[i]], id_batch,
+                                           Datum(num_groups)}));
+      RETURN_NOT_OK(agg_kernels_[i]->consume(&kernel_ctx, agg_batch));
+    }
+
+    return Status::OK();
+  }
+
+  Status OutputResult() {
+    // Finalize output
+    ArrayDataVector out_data(agg_kernels_.size() + key_field_ids_.size());
+    auto it = out_data.begin();
+
+    for (size_t i = 0; i < agg_kernels_.size(); ++i) {
+      KernelContext batch_ctx{ctx_};
+      batch_ctx.SetState(agg_states_[i].get());
+      Datum out;
+      RETURN_NOT_OK(agg_kernels_[i]->finalize(&batch_ctx, &out));
+      *it++ = out.array();
+    }
+
+    ARROW_ASSIGN_OR_RAISE(ExecBatch out_keys, grouper_->GetUniques());
+    for (const auto& key : out_keys.values) {
+      *it++ = key.array();
+    }
+
+    uint32_t num_groups = grouper_->num_groups();
+    int num_result_batches = (num_groups + output_batch_size_ - 1) / output_batch_size_;
+    outputs_[0]->InputFinished(this, num_result_batches);
+
+    for (int i = 0; i < num_result_batches; ++i) {
```

Comment:
Eventually this will need to re-seed parallelism by sumbitting the calls to `outputs_[0]->InputReceived` to the Executor